crates/allocator/src/bump.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple bump allocator.
+//!
+//! Its goal to have a much smaller footprint than the admittedly more full-featured `wee_alloc`
+//! allocator which is currently being used by ink! smart contracts.
+//!
+//! The heap which will be used by this allocator is a single page of memory, which in Wasm is
+//! 64KiB. We do not expect contracts to use more memory than this (for now), so we will throw an
+//! OOM error instead of requesting more memory.
+
+use core::alloc::{
+    GlobalAlloc,
+    Layout,
+};
+
+/// A page in Wasm is 64KiB
+const PAGE_SIZE: usize = 64 * 1024;
+
+static mut INNER: InnerAlloc = InnerAlloc::new();
+
+/// A bump allocator suitable for use in a Wasm environment.
+pub struct BumpAllocator;
+
+impl BumpAllocator {
+    /// Initialize the backing heap of the bump allocator.
+    ///
+    /// This function must only be called **once**, and it **must** be called before any
+    /// allocations are made.
+    #[inline]
+    pub fn init(&self) {
+        // SAFETY: We are in a single threaded context, so we don't have to worry about this being
+        // concurrently mutated by multiple threads.
+        unsafe { INNER.init() }
+    }
+}
+
+unsafe impl GlobalAlloc for BumpAllocator {
```

Comment:
You can override `alloc_zeroed` to just call `alloc`. We can save the zero memory operation because the new page is guaranteed to be zero initialized (on wasm).

Reply:
@athei I made the change since I believe you, but do you have any docs that mention this? I couldn't find any

Reply:
We already rely on this because cargo contracts supplies the `--zero-filled-memory` to `wasm-opt`. Without it we have a lot of zero data segments in our contract modules which bloats the binaries size.

These are the semantics for the memory as defined in the wasm module:
https://webassembly.github.io/spec/core/exec/modules.html#memories

Grow semantics:
https://webassembly.github.io/spec/core/exec/modules.html#growing-memories